File: work_project/src/test/java/automation/helpers/WebDriverExtension.java
package automation.helpers;

import automation.tests.SharedDriver;
import org.openqa.selenium.Alert;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.Collection;

/**
 * Created with IntelliJ IDEA.
 * User: Gabak Services
 * Date: 12/05/13
 * Time: 15:58
 * To change this template use File | Settings | File Templates.
 */
public class WebDriverHelpers extends SharedDriver{

    private static RemoteWebDriver Driver = SharedDriver.SharedRemoteWebDriver;
    private static Actions action = new Actions(current());
    private static Wait<WebDriver> wait;

    static {
        wait = new FluentWait<WebDriver>(current())
                .withTimeout(10, TimeUnit.SECONDS)
                .pollingEvery(1, TimeUnit.SECONDS)
                .ignoring(NoSuchElementException.class);
    }
    
    public static WebDriver current() {
        return Driver;
    }

    /*
        Custom page element finders.
     */
    public static WebElement findElementByCssSelector(String css) {
        try {
            WebElement element = current().findElement(By.cssSelector(css));
            if(!isElementDisplayed(element)) return null;
            return element;
        } catch (NoSuchElementException err) {
            throw new NoSuchElementException(err.getMessage());
        } catch (StaleElementReferenceException err) {
            throw new StaleElementReferenceException(err.getMessage());
        }
    }
    public static List<WebElement> findElementsByCssSelector(String css) {
        try {
            WebElement element = current().findElements(By.cssSelector(css));
            if(!isElementDisplayed(element)) return null;
            return element;
        } catch (NoSuchElementException err) {
            throw new NoSuchElementException(err.getMessage());
        } catch (StaleElementReferenceException err) {
            throw new StaleElementReferenceException(err.getMessage());
        }
    }

    public WebElement findElementByLinkText(String link) {
        try {
            WebElement element = current().findElements(By.linkText(link));
            if(!isElementDisplayed(element)) return null;
            return element;
        } catch (NoSuchElementException err) {
            throw new NoSuchElementException(err.getMessage());
        } catch (StaleElementReferenceException err) {
            throw new StaleElementReferenceException(err.getMessage());
        }
    }

    /*
        Custom actions.
     */
    public static void clickOn(WebElement element) {
        element.click();
    }

    public static void typeText(WebElement element, String _string) {
        if(!element.getAttribute("value").isEmpty()) {
            element.clear();
        }
        element.sendKeys(_string);
    }

    public static void navigateTo(String url) {
        current().navigate().to(url);
    }

    public static void mouseOver(WebElement element) {
        action.moveToElement(element).perform();
    }

    public static void select(WebElement element, String _string) {
        Select select = new Select(element);
        select.selectByVisibleText(_string);
    }

    public static void dragAndDrop(WebElement source, WebElement target) {
        action.dragAndDrop(source, target).perform();
    }

    public static void rightClick(WebElement element) {
        action.contextClick(element).perform();
    }

    /*
        waits for page to load and verifies page title.
     */
    public static void verifyPageTitle(String title) {
        if(!wait.until(ExpectedConditions.titleContains(title))) {
            System.err.println("Page title doesn't contain {" + title + "} but {"
                    + WEBDRIVER.getTitle() + "}");
        }
    }

    /*
        waits until element is visible on screen.
     */
    public static void waitForElement(WebElement element) {
        if(wait.until(ExpectedConditions.visibilityOf(element)) != null) return;
    }
    public static void waiter(WebElement _element) {
        final WebElement element = _element;
        wait.until(new Function<WebDriver, Object>() {
            @Override
            public Object apply(WebDriver _driver) {
                return element;
            }
        });
    }
    public static boolean isElementDisplayed(WebElement element) {
        return element.isDisplayed();
    }

    // switch driver to frame
    public static void switchToFrame(int frameId) {
        current().switchTo().frame(frameId);
    }

    /*
        alert and pop-up methods
     */
    public static Alert switchToAlert() {
        Alert alert = current().switchTo().alert();
        return alert;
    }

    public static void chooseOkOnNextConfirmation() {
        Alert alert = switchToAlert();
        alert.accept();
    }

    public static void chooseCancelOnNextConfirmation() {
        Alert alert = switchToAlert();
        alert.dismiss();
    }


}

File: dotnet/Advanced.Selenium.Tests/utils/WebDriverExtension.cs
using System;
using System.Collections.ObjectModel;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using OpenQA.Selenium.Interactions;

namespace Arena.EnhancedNews.Regression.Utils
{
    public static class WebDriverExtension
    {
        private static readonly ScreenShotRemoteWebDriver WebDriver = SharedDriver.DriverInstance.WebDriver;
        private static WebDriverWait _wait = new WebDriverWait(CurrentDriver(), SharedDriver.TimeToWait);
        private static Actions _action = new Actions(CurrentDriver());

        /// <summary>
        ///     custom driver
        /// </summary>
        /// <returns></returns>
        public static ScreenShotRemoteWebDriver CurrentDriver()
        {
            return WebDriver;
        }

        /// <summary>
        ///     custom element finders
        /// </summary>
        /// <exception cref="StaleElementReferenceException"></exception>
        public static IWebElement FindWebElementByCssSelector(string css)
        {
            try
            {
                var element = CurrentDriver().FindElementByCssSelector(css);
                return !ElementIsDisplayed(element) ? null : element;
            }
            catch (NoSuchElementException err)
            {
                
                throw new NoSuchElementException(err.Message);
            }
            catch(StaleElementReferenceException err)
            {
                throw new StaleElementReferenceException(err.Message);
            }
        }

        public static ReadOnlyCollection<IWebElement> FindWebElementsByCssSelector(string css)
        {
            return CurrentDriver().FindElementsByCssSelector(css);
        }

        public static IWebElement FindWebElementById(string id)
        {
            try
            {
                var element = CurrentDriver().FindElementById(id);
                return !ElementIsDisplayed(element) ? null : element;
            }
            catch (NoSuchElementException err)
            {

                throw new NoSuchElementException(err.Message);
            }
            catch (StaleElementReferenceException err)
            {
                throw new StaleElementReferenceException(err.Message);
            }
        }

        public static IWebElement FindWebElementByLinkText(string linkText)
        {
            try
            {
                var element = CurrentDriver().FindElementByLinkText(linkText);
                return !ElementIsDisplayed(element) ? null : element;
            }
            catch (NoSuchElementException err)
            {

                throw new NoSuchElementException(err.Message);
            }
            catch (StaleElementReferenceException err)
            {
                throw new StaleElementReferenceException(err.Message);
            }
        }

        /// <summary>
        ///     fluent wait 
        /// </summary>
        /// <returns></returns>
        public static bool ElementIsDisplayed(this IWebElement element)
        {
            return element.Displayed;
        }

        public static void Wait(string selector)
        {
            if(_wait.Until(ExpectedConditions.ElementIsVisible(By.CssSelector(selector))) != null) return;
            var exception = "Element: {" + selector + "} is not visible or loaded on " + CurrentDriver().Url;
            AssertionExtensions.ThrowShouldBeAssignableToException(exception);
        }

        /// <summary>
        ///     actions
        /// </summary>
        public static void GoTo(string uri)
        {
            CurrentDriver().Navigate().GoToUrl(uri);
        }

        public static void TypeText(this IWebElement element, string text)
        {
            if (!element.GetAttribute("value").Length.Equals(0)) element.Clear();
            element.SendKeys(text);
        }

        public static void ClickOn(this IWebElement element)
        {
            element.Click();
        }

        public static void Select(this IWebElement element, string text)
        {
            var select = new SelectElement(element);
            @select.SelectByText(text);
        }

        // to be refactored.
        public static void PressKey(this IWebElement element, string key)
        {
            if (key == null) throw new ArgumentNullException("key");
            _action.SendKeys(element, key).Perform();
        }
    }
}
